File: scripts/lib/json-schema.mjs
// ---------------------------------------------------------------------------
// JSON Schema: documented subset. Throws UnsupportedSchema on any keyword it
// does not implement, so an unvalidatable schema fails the gate.
// ---------------------------------------------------------------------------
// Extracted verbatim from scripts/kernel-validate.mjs so that the validator and
// scripts/rule-resolver.mjs check documents against JSON Schema through one
// engine, not two copies. The supported keyword set, the format checks and
// every throw are unchanged.
// ---------------------------------------------------------------------------
export class UnsupportedSchema extends Error {}

export const SUPPORTED_KEYWORDS = new Set([
  '$schema', '$id', 'id', 'title', 'description', 'examples', 'default', 'comment', '$comment',
  'type', 'properties', 'required', 'items', 'additionalProperties', 'enum', 'const',
  'pattern', 'minLength', 'maxLength', 'minItems', 'maxItems', 'uniqueItems',
  'minimum', 'maximum', 'exclusiveMinimum', 'exclusiveMaximum',
  'anyOf', 'oneOf', 'allOf', 'not', '$ref', 'definitions', '$defs', 'format',
  'if', 'then', 'else', 'propertyNames', 'contains',
]);

export function assertSupported(schema, where) {
  if (typeof schema !== 'object' || schema === null) return;
  for (const k of Object.keys(schema)) {
    if (!SUPPORTED_KEYWORDS.has(k)) {
      throw new UnsupportedSchema(`keyword "${k}" at ${where} is not implemented by this validator`);
    }
  }
}

// `format` is enforced, not merely tolerated: only the formats implemented
// here are accepted, and any other format value fails the schema up front.
// A format that was silently ignored would let a green run claim more than
// was checked.
export const FORMAT_CHECKS = {
  'date': (s) => /^\d{4}-\d{2}-\d{2}$/.test(s) && !Number.isNaN(Date.parse(s)),
  'date-time': (s) => /^\d{4}-\d{2}-\d{2}[Tt]\d{2}:\d{2}:\d{2}(\.\d+)?([Zz]|[+-]\d{2}:\d{2})$/.test(s) && !Number.isNaN(Date.parse(s)),
  'uri': (s) => { try { new URL(s); return true; } catch { return false; } },
};

// Walk the whole schema tree once, before any data is validated. Per-node
// assertion during validation only reaches the branches the data happens to
// visit; an unsupported keyword in an optional, never-taken branch would
// otherwise pass silently.
const SCHEMA_MAP_KEYWORDS = ['properties', 'definitions', '$defs'];
const SCHEMA_CHILD_KEYWORDS = ['items', 'additionalProperties', 'not', 'if', 'then', 'else', 'propertyNames', 'contains'];
const SCHEMA_LIST_KEYWORDS = ['anyOf', 'oneOf', 'allOf'];
export function assertSupportedDeep(schema, where) {
  if (typeof schema !== 'object' || schema === null || Array.isArray(schema)) return;
  assertSupported(schema, where);
  if (schema.format != null && !FORMAT_CHECKS[schema.format]) {
    throw new UnsupportedSchema(`format "${schema.format}" at ${where} is not implemented by this validator`);
  }
  for (const k of SCHEMA_MAP_KEYWORDS) {
    if (schema[k] && typeof schema[k] === 'object') {
      for (const [name, sub] of Object.entries(schema[k])) assertSupportedDeep(sub, `${where}/${k}/${name}`);
    }
  }
  for (const k of SCHEMA_CHILD_KEYWORDS) {
    if (typeof schema[k] === 'object' && schema[k] !== null) assertSupportedDeep(schema[k], `${where}/${k}`);
  }
  for (const k of SCHEMA_LIST_KEYWORDS) {
    if (Array.isArray(schema[k])) schema[k].forEach((sub, i) => assertSupportedDeep(sub, `${where}/${k}/${i}`));
  }
}

function resolveRef(ref, root) {
  if (!ref.startsWith('#/')) throw new UnsupportedSchema(`external $ref "${ref}"`);
  let node = root;
  for (const seg of ref.slice(2).split('/')) {
    node = node?.[seg.replace(/~1/g, '/').replace(/~0/g, '~')];
    if (node === undefined) throw new UnsupportedSchema(`unresolvable $ref "${ref}"`);
  }
  return node;
}

function typeOf(v) {
  if (v === null) return 'null';
  if (Array.isArray(v)) return 'array';
  if (Number.isInteger(v)) return 'integer';
  return typeof v;
}

export function validate(data, schema, root, ptr, errs) {
  assertSupported(schema, ptr || '/');
  if (schema.$ref) return validate(data, resolveRef(schema.$ref, root), root, ptr, errs);

  if (schema.type) {
    const types = Array.isArray(schema.type) ? schema.type : [schema.type];
    const actual = typeOf(data);
    const okType = types.some((t) => t === actual || (t === 'number' && actual === 'integer'));
    if (!okType) { errs.push(`${ptr || '/'}: expected ${types.join('|')}, got ${actual}`); return; }
  }
  if (schema.enum && !schema.enum.some((e) => JSON.stringify(e) === JSON.stringify(data))) {
    errs.push(`${ptr || '/'}: value ${JSON.stringify(data)} not in enum`);
  }
  if ('const' in schema && JSON.stringify(schema.const) !== JSON.stringify(data)) {
    errs.push(`${ptr || '/'}: value must equal ${JSON.stringify(schema.const)}`);
  }
  if (typeof data === 'string') {
    if (schema.pattern && !new RegExp(schema.pattern).test(data)) errs.push(`${ptr}: does not match /${schema.pattern}/`);
    if (schema.minLength != null && data.length < schema.minLength) errs.push(`${ptr}: shorter than ${schema.minLength}`);
    if (schema.maxLength != null && data.length > schema.maxLength) errs.push(`${ptr}: longer than ${schema.maxLength}`);
    if (schema.format != null) {
      const check = FORMAT_CHECKS[schema.format];
      if (!check) throw new UnsupportedSchema(`format "${schema.format}" at ${ptr || '/'} is not implemented by this validator`);
      if (!check(data)) errs.push(`${ptr}: does not satisfy format "${schema.format}"`);
    }
  }
  if (typeof data === 'number') {
    if (schema.minimum != null && data < schema.minimum) errs.push(`${ptr}: below minimum`);
    if (schema.maximum != null && data > schema.maximum) errs.push(`${ptr}: above maximum`);
    if (schema.exclusiveMinimum != null && data <= schema.exclusiveMinimum) errs.push(`${ptr}: at/below exclusiveMinimum`);
    if (schema.exclusiveMaximum != null && data >= schema.exclusiveMaximum) errs.push(`${ptr}: at/above exclusiveMaximum`);
  }
  if (Array.isArray(data)) {
    if (schema.minItems != null && data.length < schema.minItems) errs.push(`${ptr}: fewer than ${schema.minItems} items`);
    if (schema.maxItems != null && data.length > schema.maxItems) errs.push(`${ptr}: more than ${schema.maxItems} items`);
    if (schema.uniqueItems) {
      const seen = new Set(data.map((d) => JSON.stringify(d)));
      if (seen.size !== data.length) errs.push(`${ptr}: items are not unique`);
    }
    if (schema.items) data.forEach((d, i) => validate(d, schema.items, root, `${ptr}/${i}`, errs));
    if (schema.contains) {
      const hit = data.some((d) => { const e = []; validate(d, schema.contains, root, ptr, e); return e.length === 0; });
      if (!hit) errs.push(`${ptr || '/'}: no item matches "contains"`);
    }
  }
  if (data && typeof data === 'object' && !Array.isArray(data)) {
    for (const req of schema.required || []) {
      if (!(req in data) || data[req] === undefined) errs.push(`${ptr || ''}/${req}: required property missing`);
    }
    if (schema.propertyNames) {
      for (const k of Object.keys(data)) validate(k, schema.propertyNames, root, `${ptr}/${k}<name>`, errs);
    }
    for (const [k, v] of Object.entries(data)) {
      const sub = schema.properties?.[k];
      if (sub) validate(v, sub, root, `${ptr}/${k}`, errs);
      else if (schema.additionalProperties === false) errs.push(`${ptr}/${k}: additional property not allowed`);
      else if (typeof schema.additionalProperties === 'object') validate(v, schema.additionalProperties, root, `${ptr}/${k}`, errs);
    }
  }
  for (const kw of ['allOf']) if (schema[kw]) schema[kw].forEach((s, i) => validate(data, s, root, `${ptr}/allOf/${i}`, errs));
  for (const kw of ['anyOf', 'oneOf']) {
    if (!schema[kw]) continue;
    const passing = schema[kw].filter((s) => { const e = []; validate(data, s, root, ptr, e); return e.length === 0; }).length;
    if (kw === 'anyOf' && passing === 0) errs.push(`${ptr || '/'}: matches none of anyOf`);
    if (kw === 'oneOf' && passing !== 1) errs.push(`${ptr || '/'}: matches ${passing} of oneOf (expected exactly 1)`);
  }
  if (schema.not) { const e = []; validate(data, schema.not, root, ptr, e); if (e.length === 0) errs.push(`${ptr || '/'}: must not match "not" schema`); }
  if (schema.if) {
    const probe = [];
    validate(data, schema.if, root, ptr, probe);
    const branch = probe.length === 0 ? schema.then : schema.else;
    if (branch) validate(data, branch, root, ptr, errs);
  }
}

File: meridian-app/src/source_format/json_schema.rs
//! Strict Draft 7 JSON Schema adapter.
//!
//! `jsonschema` compiles every accepted schema. An allowlist and compatibility
//! evaluator then preserve the narrower observable contract of
//! `scripts/lib/json-schema.mjs`, including its diagnostic text.

use std::{collections::HashSet, fmt};

use fancy_regex::Regex;
use serde_json::Value;

pub const SUPPORTED_KEYWORDS: &[&str] = &[
    "$schema",
    "$id",
    "id",
    "title",
    "description",
    "examples",
    "default",
    "comment",
    "$comment",
    "type",
    "properties",
    "required",
    "items",
    "additionalProperties",
    "enum",
    "const",
    "pattern",
    "minLength",
    "maxLength",
    "minItems",
    "maxItems",
    "uniqueItems",
    "minimum",
    "maximum",
    "exclusiveMinimum",
    "exclusiveMaximum",
    "anyOf",
    "oneOf",
    "allOf",
    "not",
    "$ref",
    "definitions",
    "$defs",
    "format",
    "if",
    "then",
    "else",
    "propertyNames",
    "contains",
];

pub const SUPPORTED_FORMATS: &[&str] = &["date", "date-time", "uri"];

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct UnsupportedSchema(pub String);

impl fmt::Display for UnsupportedSchema {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(&self.0)
    }
}

impl std::error::Error for UnsupportedSchema {}

/// Reject unsupported surface everywhere, compile as Draft 7, then return
/// diagnostics matching the Node.js reference evaluator.
pub fn validate(data: &Value, schema: &Value) -> Result<Vec<String>, UnsupportedSchema> {
    assert_supported_deep(schema, "/")?;
    validate_refs(schema, schema)?;
    let validator = jsonschema::draft7::new(schema)
        .map_err(|error| UnsupportedSchema(format!("invalid Draft 7 schema: {error}")))?;
    // Compilation is the stable library boundary. Running it also exercises
    // the compiled graph; compatibility diagnostics below remain the public
    // result because library wording is not cross-language stable.
    let _library_verdict = validator.is_valid(data);
    let mut errors = Vec::new();
    validate_node(data, schema, schema, "", &mut errors)?;
    Ok(errors)
}

pub fn assert_supported_deep(schema: &Value, where_: &str) -> Result<(), UnsupportedSchema> {
    let Some(object) = schema.as_object() else {
        return Ok(());
    };
    for key in object.keys() {
        if !SUPPORTED_KEYWORDS.contains(&key.as_str()) {
            return Err(UnsupportedSchema(format!(
                "keyword \"{key}\" at {where_} is not implemented by this validator"
            )));
        }
    }
    if let Some(format) = object.get("format").and_then(Value::as_str) {
        if !SUPPORTED_FORMATS.contains(&format) {
            return Err(UnsupportedSchema(format!(
                "format \"{format}\" at {where_} is not implemented by this validator"
            )));
        }
    }
    for keyword in ["properties", "definitions", "$defs"] {
        if let Some(children) = object.get(keyword).and_then(Value::as_object) {
            for (name, child) in children {
                assert_supported_deep(child, &format!("{where_}/{keyword}/{name}"))?;
            }
        }
    }
    for keyword in [
        "items",
        "additionalProperties",
        "not",
        "if",
        "then",
        "else",
        "propertyNames",
        "contains",
    ] {
        if let Some(child) = object.get(keyword).filter(|value| value.is_object()) {
            assert_supported_deep(child, &format!("{where_}/{keyword}"))?;
        }
    }
    for keyword in ["anyOf", "oneOf", "allOf"] {
        if let Some(children) = object.get(keyword).and_then(Value::as_array) {
            for (index, child) in children.iter().enumerate() {
                assert_supported_deep(child, &format!("{where_}/{keyword}/{index}"))?;
            }
        }
    }
    Ok(())
}

fn validate_refs(schema: &Value, root: &Value) -> Result<(), UnsupportedSchema> {
    let Some(object) = schema.as_object() else {
        return Ok(());
    };
    if let Some(reference) = object.get("$ref").and_then(Value::as_str) {
        resolve_ref(reference, root)?;
    }
    for keyword in ["properties", "definitions", "$defs"] {
        if let Some(children) = object.get(keyword).and_then(Value::as_object) {
            for child in children.values() {
                validate_refs(child, root)?;
            }
        }
    }
    for keyword in [
        "items",
        "additionalProperties",
        "not",
        "if",
        "then",
        "else",
        "propertyNames",
        "contains",
    ] {
        if let Some(child) = object.get(keyword).filter(|value| value.is_object()) {
            validate_refs(child, root)?;
        }
    }
    for keyword in ["anyOf", "oneOf", "allOf"] {
        if let Some(children) = object.get(keyword).and_then(Value::as_array) {
            for child in children {
                validate_refs(child, root)?;
            }
        }
    }
    Ok(())
}

fn resolve_ref<'a>(reference: &str, root: &'a Value) -> Result<&'a Value, UnsupportedSchema> {
    let Some(pointer) = reference.strip_prefix("#/") else {
        return Err(UnsupportedSchema(format!("external $ref \"{reference}\"")));
    };
    let mut node = root;
    for segment in pointer.split('/') {
        let key = segment.replace("~1", "/").replace("~0", "~");
        let Some(next) = node.get(&key) else {
            return Err(UnsupportedSchema(format!(
                "unresolvable $ref \"{reference}\""
            )));
        };
        node = next;
    }
    Ok(node)
}

fn validate_node(
    data: &Value,
    schema: &Value,
    root: &Value,
    pointer: &str,
    errors: &mut Vec<String>,
) -> Result<(), UnsupportedSchema> {
    let Some(object) = schema.as_object() else {
        return Ok(());
    };
    if let Some(reference) = object.get("$ref").and_then(Value::as_str) {
        return validate_node(data, resolve_ref(reference, root)?, root, pointer, errors);
    }
    if let Some(types) = object.get("type") {
        let expected: Vec<&str> = match types {
            Value::String(value) => vec![value],
            Value::Array(values) => values.iter().filter_map(Value::as_str).collect(),
            _ => Vec::new(),
        };
        let actual = type_of(data);
        let matches = expected
            .iter()
            .any(|kind| *kind == actual || (*kind == "number" && actual == "integer"));
        if !matches {
            errors.push(format!(
                "{}: expected {}, got {actual}",
                display_pointer(pointer),
                expected.join("|")
            ));
            return Ok(());
        }
    }
    if let Some(values) = object.get("enum").and_then(Value::as_array) {
        if !values.iter().any(|value| json_equal(value, data)) {
            errors.push(format!(
                "{}: value {} not in enum",
                display_pointer(pointer),
                json_text(data)
            ));
        }
    }
    if let Some(expected) = object.get("const") {
        if !json_equal(expected, data) {
            errors.push(format!(
                "{}: value must equal {}",
                display_pointer(pointer),
                json_text(expected)
            ));
        }
    }
    if let Some(text) = data.as_str() {
        validate_string(text, object, pointer, errors)?;
    }
    if let Some(number) = data.as_f64() {
        validate_number(number, object, pointer, errors);
    }
    if let Some(values) = data.as_array() {
        validate_array(values, object, root, pointer, errors)?;
    }
    if let Some(values) = data.as_object() {
        validate_object(values, object, root, pointer, errors)?;
    }

    if let Some(schemas) = object.get("allOf").and_then(Value::as_array) {
        for (index, child) in schemas.iter().enumerate() {
            validate_node(
                data,
                child,
                root,
                &format!("{pointer}/allOf/{index}"),
                errors,
            )?;
        }
    }
    for keyword in ["anyOf", "oneOf"] {
        let Some(schemas) = object.get(keyword).and_then(Value::as_array) else {
            continue;
        };
        let mut passing = 0;
        for child in schemas {
            let mut probe = Vec::new();
            validate_node(data, child, root, pointer, &mut probe)?;
            if probe.is_empty() {
                passing += 1;
            }
        }
        if keyword == "anyOf" && passing == 0 {
            errors.push(format!(
                "{}: matches none of anyOf",
                display_pointer(pointer)
            ));
        }
        if keyword == "oneOf" && passing != 1 {
            errors.push(format!(
                "{}: matches {passing} of oneOf (expected exactly 1)",
                display_pointer(pointer)
            ));
        }
    }
    if let Some(child) = object.get("not") {
        let mut probe = Vec::new();
        validate_node(data, child, root, pointer, &mut probe)?;
        if probe.is_empty() {
            errors.push(format!(
                "{}: must not match \"not\" schema",
                display_pointer(pointer)
            ));
        }
    }
    if let Some(condition) = object.get("if") {
        let mut probe = Vec::new();
        validate_node(data, condition, root, pointer, &mut probe)?;
        let branch = if probe.is_empty() {
            object.get("then")
        } else {
            object.get("else")
        };
        if let Some(branch) = branch {
            validate_node(data, branch, root, pointer, errors)?;
        }
    }
    Ok(())
}

fn validate_string(
    text: &str,
    schema: &serde_json::Map<String, Value>,
    pointer: &str,
    errors: &mut Vec<String>,
) -> Result<(), UnsupportedSchema> {
    if let Some(pattern) = schema.get("pattern").and_then(Value::as_str) {
        let regex = Regex::new(pattern).map_err(|error| {
            UnsupportedSchema(format!("invalid pattern \"{pattern}\": {error}"))
        })?;
        if !regex
            .is_match(text)
            .map_err(|error| UnsupportedSchema(format!("pattern evaluation failed: {error}")))?
        {
            errors.push(format!("{pointer}: does not match /{pattern}/"));
        }
    }
    let length = text.encode_utf16().count() as u64;
    if let Some(minimum) = schema.get("minLength").and_then(Value::as_u64) {
        if length < minimum {
            errors.push(format!("{pointer}: shorter than {minimum}"));
        }
    }
    if let Some(maximum) = schema.get("maxLength").and_then(Value::as_u64) {
        if length > maximum {
            errors.push(format!("{pointer}: longer than {maximum}"));
        }
    }
    if let Some(format) = schema.get("format").and_then(Value::as_str) {
        if !format_valid(format, text) {
            errors.push(format!("{pointer}: does not satisfy format \"{format}\""));
        }
    }
    Ok(())
}

fn validate_number(
    number: f64,
    schema: &serde_json::Map<String, Value>,
    pointer: &str,
    errors: &mut Vec<String>,
) {
    if schema
        .get("minimum")
        .and_then(Value::as_f64)
        .is_some_and(|limit| number < limit)
    {
        errors.push(format!("{pointer}: below minimum"));
    }
    if schema
        .get("maximum")
        .and_then(Value::as_f64)
        .is_some_and(|limit| number > limit)
    {
        errors.push(format!("{pointer}: above maximum"));
    }
    if schema
        .get("exclusiveMinimum")
        .and_then(Value::as_f64)
        .is_some_and(|limit| number <= limit)
    {
        errors.push(format!("{pointer}: at/below exclusiveMinimum"));
    }
    if schema
        .get("exclusiveMaximum")
        .and_then(Value::as_f64)
        .is_some_and(|limit| number >= limit)
    {
        errors.push(format!("{pointer}: at/above exclusiveMaximum"));
    }
}

fn validate_array(
    values: &[Value],
    schema: &serde_json::Map<String, Value>,
    root: &Value,
    pointer: &str,
    errors: &mut Vec<String>,
) -> Result<(), UnsupportedSchema> {
    if schema
        .get("minItems")
        .and_then(Value::as_u64)
        .is_some_and(|limit| values.len() < limit as usize)
    {
        errors.push(format!(
            "{pointer}: fewer than {} items",
            schema["minItems"]
        ));
    }
    if schema
        .get("maxItems")
        .and_then(Value::as_u64)
        .is_some_and(|limit| values.len() > limit as usize)
    {
        errors.push(format!("{pointer}: more than {} items", schema["maxItems"]));
    }
    if schema.get("uniqueItems").and_then(Value::as_bool) == Some(true) {
        let unique: HashSet<String> = values.iter().map(json_text).collect();
        if unique.len() != values.len() {
            errors.push(format!("{pointer}: items are not unique"));
        }
    }
    if let Some(item_schema) = schema.get("items") {
        for (index, value) in values.iter().enumerate() {
            validate_node(
                value,
                item_schema,
                root,
                &format!("{pointer}/{index}"),
                errors,
            )?;
        }
    }
    if let Some(contains) = schema.get("contains") {
        let mut hit = false;
        for value in values {
            let mut probe = Vec::new();
            validate_node(value, contains, root, pointer, &mut probe)?;
            if probe.is_empty() {
                hit = true;
                break;
            }
        }
        if !hit {
            errors.push(format!(
                "{}: no item matches \"contains\"",
                display_pointer(pointer)
            ));
        }
    }
    Ok(())
}

fn validate_object(
    values: &serde_json::Map<String, Value>,
    schema: &serde_json::Map<String, Value>,
    root: &Value,
    pointer: &str,
    errors: &mut Vec<String>,
) -> Result<(), UnsupportedSchema> {
    if let Some(required) = schema.get("required").and_then(Value::as_array) {
        for name in required.iter().filter_map(Value::as_str) {
            if !values.contains_key(name) {
                errors.push(format!("{pointer}/{name}: required property missing"));
            }
        }
    }
    if let Some(name_schema) = schema.get("propertyNames") {
        for name in values.keys() {
            validate_node(
                &Value::String(name.clone()),
                name_schema,
                root,
                &format!("{pointer}/{name}<name>"),
                errors,
            )?;
        }
    }
    let properties = schema.get("properties").and_then(Value::as_object);
    for (name, value) in values {
        if let Some(child) = properties.and_then(|map| map.get(name)) {
            validate_node(value, child, root, &format!("{pointer}/{name}"), errors)?;
        } else if schema.get("additionalProperties") == Some(&Value::Bool(false)) {
            errors.push(format!("{pointer}/{name}: additional property not allowed"));
        } else if let Some(child) = schema
            .get("additionalProperties")
            .filter(|value| value.is_object())
        {
            validate_node(value, child, root, &format!("{pointer}/{name}"), errors)?;
        }
    }
    Ok(())
}

fn format_valid(format: &str, value: &str) -> bool {
    match format {
        "date" => {
            let parts: Vec<_> = value.split('-').collect();
            parts.len() == 3
                && parts[0].len() == 4
                && parts[1].len() == 2
                && parts[2].len() == 2
                && parts
                    .iter()
                    .all(|part| part.bytes().all(|byte| byte.is_ascii_digit()))
                && parts[1]
                    .parse::<u8>()
                    .is_ok_and(|month| (1..=12).contains(&month))
                && parts[2]
                    .parse::<u8>()
                    .is_ok_and(|day| (1..=31).contains(&day))
        }
        "date-time" => date_time_valid(value),
        "uri" => value.split_once(':').is_some_and(|(scheme, rest)| {
            !scheme.is_empty()
                && !rest.is_empty()
                && scheme.bytes().enumerate().all(|(index, byte)| {
                    if index == 0 {
                        byte.is_ascii_alphabetic()
                    } else {
                        byte.is_ascii_alphanumeric() || b"+.-".contains(&byte)
                    }
                })
        }),
        _ => false,
    }
}

fn date_time_valid(value: &str) -> bool {
    let regex = Regex::new(
        r"^\d{4}-(\d{2})-(\d{2})[Tt](\d{2}):(\d{2}):(\d{2})(\.\d+)?([Zz]|[+-](\d{2}):(\d{2}))$",
    )
    .expect("static date-time regex");
    let Ok(Some(captures)) = regex.captures(value) else {
        return false;
    };
    let number = |index| {
        captures
            .get(index)
            .and_then(|capture| capture.as_str().parse::<u8>().ok())
    };
    number(1).is_some_and(|month| (1..=12).contains(&month))
        && number(2).is_some_and(|day| (1..=31).contains(&day))
        && number(3).is_some_and(|hour| hour <= 23)
        && number(4).is_some_and(|minute| minute <= 59)
        && number(5).is_some_and(|second| second <= 59)
        && number(8).is_none_or(|hour| hour <= 23)
        && number(9).is_none_or(|minute| minute <= 59)
}

fn type_of(value: &Value) -> &'static str {
    match value {
        Value::Null => "null",
        Value::Bool(_) => "boolean",
        Value::Number(number) if number.is_i64() || number.is_u64() => "integer",
        Value::Number(_) => "number",
        Value::String(_) => "string",
        Value::Array(_) => "array",
        Value::Object(_) => "object",
    }
}

fn display_pointer(pointer: &str) -> &str {
    if pointer.is_empty() {
        "/"
    } else {
        pointer
    }
}
fn json_text(value: &Value) -> String {
    serde_json::to_string(value).expect("JSON value serializes")
}
fn json_equal(left: &Value, right: &Value) -> bool {
    json_text(left) == json_text(right)
}

#[cfg(test)]
mod tests {
    use super::{assert_supported_deep, validate};
    use serde_json::json;

    #[test]
    fn validates_supported_schema_with_reference_diagnostics() {
        let schema = json!({"type":"object","required":["id"],"properties":{"id":{"type":"string","minLength":2}},"additionalProperties":false});
        assert_eq!(
            validate(&json!({"id":"x","extra":true}), &schema).unwrap(),
            vec![
                "/extra: additional property not allowed",
                "/id: shorter than 2"
            ]
        );
    }

    #[test]
    fn rejects_unsupported_keywords_in_unvisited_branches() {
        let schema = json!({"properties":{"optional":{"type":"string","minProperties":1}}});
        assert!(assert_supported_deep(&schema, "/").is_err());
    }

    #[test]
    fn rejects_unknown_formats_and_external_references() {
        assert!(validate(&json!("x"), &json!({"type":"string","format":"email"})).is_err());
        assert!(validate(&json!(1), &json!({"$ref":"https://example.test/schema"})).is_err());
    }
}
